r/R/convert-array-stream.R

```diff
@@ -62,26 +68,52 @@ convert_array_stream <- function(array_stream, to = NULL, size = NULL, n = Inf)
     )
   }
 
+  # Collect all batches into a list()
+  batches <- collect_array_stream(
+    array_stream,
+    n,
+    schema = schema,
+    validate = FALSE
+  )
+
+  # If there is exactly one batch, use convert_array()
+  if (length(batches) == 1L) {
+    return(.Call(nanoarrow_c_convert_array, batches[[1]], to))
+  }
+
+  # Otherwise, compute the final size, create another array stream,
+  # and call convert_array_stream() with a known size.
+  lengths <- vapply(
+    batches,
+    # Use custom accessor because array$length in a loop is slow
+    function(array) {
+      .Call(nanoarrow_c_array_proxy, array, NULL, FALSE)$length
+    },
+    double(1)
+  )
+  basic_stream <- .Call(nanoarrow_c_basic_array_stream, batches, schema, FALSE)
+  convert_array_stream(basic_stream, to = to, size = sum(lengths))
+}
+
+#' @rdname convert_array_stream
+#' @export
+collect_array_stream <- function(array_stream, n = Inf, schema = NULL,
+                                 validate = TRUE) {
+  stopifnot(
+    inherits(array_stream, "nanoarrow_array_stream")
+  )
+
+  if (is.null(schema)) {
+    schema <- .Call(nanoarrow_c_array_stream_get_schema, array_stream)
+  }
+
   batches <- vector("list", 1024L)
   n_batches <- 0L
   get_next <- array_stream$get_next
-  while (!is.null(array <- get_next(schema, validate = FALSE)) && (n_batches < n)) {
+  while (!is.null(array <- get_next(schema, validate = validate)) && (n_batches < n)) {
```

Comment:
I prefer a `break` inside the loop for the `n_batches < n` case, it also makes it very clear when the check occurs.